Use FPTrunc and FPExt in vectorized code instead of failing conversions.

csvtuda opened the pull request:
This PR allows gaussian leaf vectorization patterns to extend or truncate the float bit width of their input vectors. It does so by applying `std::FPTruncOp` and `std::FPExtOp` to vectors of floats as needed.

Although these two operations are explicitly marked to not be applicable to vectors in the [documentation](https://mlir.llvm.org/docs/Dialects/Standard/#stdfpext-mlirfpextop), this is merely due to the documentation being outdated, as discussed [here](https://llvm.discourse.group/t/vector-of-doubles-to-vector-of-floats-type-conversion/3619). I have verified this during testing of the SLP vectorizer, where `vector<8xf64>` have successfully been shortened to `vector<8xf32>`.

sommerlukas reviewed (changes requested):
Thanks for investigating this and the update of the patterns. Looks good so far!

Please also update the check for vectorizability for the Gaussian leaf [here](https://github.com/esa-tu-darmstadt/spn-compiler/blob/0a60a70be2a46c5f8fec87fcdd22f6328c065613/mlir/lib/Dialect/LoSPN/LoSPNOps.cpp#L374) to reflect the new information/pattern.

And, if possible, please also add a test that the vectorization is actually performed if the Gaussian leaf required `fptrunc/fpext`.

csvtuda commented:
Thanks for the feedback.
> Please also update the check for vectorizability for the Gaussian leaf here to reflect the new information/pattern.

Whoops I totally forgot about that .cpp file. Addressed in 3992490c94a08354f19be34678f1c336d0f08f1e.

> And, if possible, please also add a test that the vectorization is actually performed if the Gaussian leaf required fptrunc/fpext.

I added two test cases in 9a4a0859ca5ee9fed59cc9f8e672a175e7b48cac: one where the input feature vector is extended and one where it is truncated.

sommerlukas reviewed (approved):
Thanks for the update, looks good to me.
